pkg/remotes/docker/artifacts.go

```diff
@@ -0,0 +1,266 @@
+		return nil, errors.New("not implemented")
+	}
+
+	req.header.Set("Content-Type", desc.MediaType)
+
+	pr, pw := io.Pipe()
+	respC := make(chan response, 1)
+	body := ioutil.NopCloser(pr)
+
+	req.body = func() (io.ReadCloser, error) {
+		if body == nil {
+			return nil, errors.New("retry request, cannot reset the stream")
+		}
+
+		readout := body
+		body = nil
+		return readout, nil
+	}
+	req.size = desc.Size
+
+	go func() {
+		defer close(respC)
+		resp, err := req.doWithRetries(ctx, nil)
+		if err != nil {
+			respC <- response{err: err}
+			pr.CloseWithError(err)
+			return
+		}
+
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusCreated, http.StatusNoContent:
+		default:
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			pr.CloseWithError(err)
+		}
+		respC <- response{Response: resp}
+	}()
+
+	return &artifactsManifest{
+		ref:       d.reference,
+		expected:  desc.Digest,
+		pipe:      pw,
+		responseC: respC,
+		tracker:   d.tracker,
+	}, nil
+}
+
+func (d *dockerDiscoverer) CheckManifest(ctx context.Context, host docker.RegistryHost, desc ocispec.Descriptor) error {
+	// Check if the manifest exists
+	req := d.request(host, http.MethodHead, "manifests", desc.Digest.String())
+	if req == nil {
+		return errors.New("not implemented")
+	}
+
+	req.header.Set("Accept", desc.MediaType)
+
+	resp, err := req.doWithRetries(ctx, nil)
+	if err != nil {
+		if !errors.Is(err, docker.ErrInvalidAuthorization) {
+			return err
+		}
+	} else {
+		if resp.StatusCode == http.StatusOK && resp.Header.Get("Docker-Content-Digest") == desc.Digest.String() {
+			d.tracker.SetStatus(d.reference, docker.Status{
+				Committed: true,
+				Status: content.Status{
+					Ref:    d.reference,
+					Total:  desc.Size,
+					Offset: desc.Size,
+				}})
+
+			return errdefs.ErrAlreadyExists
+		} else if resp.StatusCode != http.StatusNotFound {
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			resp.Body.Close()
+			return err
+		}
+	}
+
+	return nil
+}
+
+type response struct {
+	*http.Response
+	err error
+}
+
+type artifactsManifest struct {
+	ref       string
+	expected  digest.Digest
+	pipe      *io.PipeWriter
+	tracker   docker.StatusTracker
+	responseC <-chan response
+}
+
+func (pw *artifactsManifest) Write(p []byte) (n int, err error) {
+	status, err := pw.tracker.GetStatus(pw.ref)
+	if err != nil {
+		return n, err
+	}
+	n, err = pw.pipe.Write(p)
+	status.Offset += int64(n)
+	status.UpdatedAt = time.Now()
+	pw.tracker.SetStatus(pw.ref, status)
+	return
+}
+
+func (pw *artifactsManifest) Close() error {
+	return pw.pipe.Close()
+}
+
+func (pw *artifactsManifest) Status() (content.Status, error) {
+	status, err := pw.tracker.GetStatus(pw.ref)
+	if err != nil {
+		return content.Status{}, err
+	}
+	return status.Status, nil
+
+}
+
+func (pw *artifactsManifest) Digest() digest.Digest {
+	return pw.expected
+}
+
+func (pw *artifactsManifest) Commit(ctx context.Context, size int64, expected digest.Digest, opts ...content.Opt) error {
```

Comment:
where is the method used?

Reply:
This will eventually get called by the content.Writer